calico/felix/devices.py

```diff
@@ -264,27 +254,25 @@ def interface_up(if_name):
     """
     Checks whether a given interface is up.
 
-    Check this by examining the interface flags and looking for whether the
-    IFF_UP flag has been set. IFF_UP is the flag 0x01, so read in the flags
-    (represented by a hexadecimal integer) and check if the 1 bit is set.
+    Check this by examining the operstate of the interface, which is the
+    highest level "is it ready to work with" flag.
 
     :param str if_name: Interface name
     :returns: True if interface up, False if down or cannot detect
     """
-    flags_file = '/sys/class/net/%s/flags' % if_name
-
+    operstate_filename = '/sys/class/net/%s/operstate' % if_name
     try:
-        with open(flags_file, 'r') as f:
-            flags = f.read().strip()
-            _log.debug("Interface %s has flags %s", if_name, flags)
+        with open(operstate_filename, 'r') as f:
+            oper_state = f.read().strip()
     except IOError as e:
         # If we fail to check that the interface is up, then it has probably
         # gone under our feet or is flapping.
-        _log.warning("Cannot check flags for interface %s (%s) - assume "
-                     "down: %r.", if_name, flags_file, e)
+        _log.warning("Failed to read state of interface %s (%s) - assume "
+                     "down/absent: %r.", if_name, operstate_filename, e)
         return False
-
-    return bool(int(flags, 16) & 1)
+    else:
+        _log.debug("Interface %s has state %s", if_name, oper_state)
+    return oper_state == "up"
```

Comment:
Doh. Yes, you are indeed correct. I have no idea how I managed to miss that, given that it's on the next line to the one I was looking at. I withdraw my objection and shall return to my slumber.